File: lca/src/main/java/com/dreampany/lca/data/model/Quote.java
package com.dreampany.lca.data.model;

import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Index;
import android.os.Parcel;
import androidx.annotation.NonNull;
import com.dreampany.frame.data.model.Base;
import org.apache.commons.lang3.builder.ReflectionToStringBuilder;

/**
 * Created by Roman on 1/14/2019
 * Copyright (c) 2019 Dreampany. All rights reserved.
 * dreampanymail@gmail.com
 * Last modified $file.lastModified
 */

@Entity(indices = {@Index(value = {"coinId", "currency"}, unique = true)},
        primaryKeys = {"coinId", "currency"})
public class Quote extends Base {

    private long coinId; // id of coin
    @NonNull
    private Currency currency;
    private double price;
    private double dayVolume;
    private double marketCap;
    private double hourChange;
    private double dayChange;
    private double weekChange;
    private String lastUpdated;

    @Ignore
    public Quote() {

    }

    public Quote(long id) {
        this.id = id;
    }

    @Ignore
    private Quote(Parcel in) {
        super(in);
        coinId = in.readLong();
        currency = in.readParcelable(Currency.class.getClassLoader());
        price = in.readDouble();
        dayVolume = in.readDouble();
        marketCap = in.readDouble();
        hourChange = in.readDouble();
        dayChange = in.readDouble();
        weekChange = in.readDouble();
        lastUpdated = in.readString();
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeLong(coinId);
        dest.writeParcelable(currency, flags);
        dest.writeDouble(price);
        dest.writeDouble(dayVolume);
        dest.writeDouble(marketCap);
        dest.writeDouble(hourChange);
        dest.writeDouble(dayChange);
        dest.writeDouble(weekChange);
        dest.writeString(lastUpdated);
    }

    public static final Creator<Quote> CREATOR = new Creator<Quote>() {
        @Override
        public Quote createFromParcel(Parcel in) {
            return new Quote(in);
        }

        @Override
        public Quote[] newArray(int size) {
            return new Quote[size];
        }
    };

    @NonNull
    @Override
    public String toString() {
        return ReflectionToStringBuilder.toString(this);
    }

    public void setCoinId(long coinId) {
        this.coinId = coinId;
    }

    public void setCurrency(Currency currency) {
        this.currency = currency;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public void setDayVolume(double dayVolume) {
        this.dayVolume = dayVolume;
    }

    public void setMarketCap(double marketCap) {
        this.marketCap = marketCap;
    }

    public void setHourChange(double hourChange) {
        this.hourChange = hourChange;
    }

    public void setDayChange(double dayChange) {
        this.dayChange = dayChange;
    }

    public void setWeekChange(double weekChange) {
        this.weekChange = weekChange;
    }

    public void setLastUpdated(String lastUpdated) {
        this.lastUpdated = lastUpdated;
    }

    public long getCoinId() {
        return coinId;
    }

    public Currency getCurrency() {
        return currency;
    }

    public double getPrice() {
        return price;
    }

    public double getDayVolume() {
        return dayVolume;
    }

    public double getMarketCap() {
        return marketCap;
    }

    public double getHourChange() {
        return hourChange;
    }

    public double getDayChange() {
        return dayChange;
    }

    public double getWeekChange() {
        return weekChange;
    }

    public String getLastUpdated() {
        return lastUpdated;
    }
}

File: tools/src/main/kotlin/com/dreampany/tools/data/model/Quote.kt
package com.dreampany.tools.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import com.dreampany.framework.data.model.Base
import com.dreampany.framework.util.TimeUtilKt
import com.dreampany.tools.data.enums.Currency
import com.dreampany.tools.misc.Constants
import com.google.common.base.Objects
import com.google.firebase.database.IgnoreExtraProperties
import com.google.firebase.database.PropertyName
import kotlinx.android.parcel.Parcelize

/**
 * Created by roman on 2019-10-01
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Parcelize
@IgnoreExtraProperties
@Entity(
    indices = [Index(
        value = [Constants.Database.Crypto.ID, Constants.Database.Crypto.CURRENCY],
        unique = true
    )],
    primaryKeys = [Constants.Database.Crypto.ID, Constants.Database.Crypto.CURRENCY]
)
data class Quote(
    override var time: Long = Constants.Default.LONG,
    override var id: String = Constants.Default.STRING,
    var currency: Currency = Currency.USD,
    var price: Double = Constants.Default.DOUBLE,
    @ColumnInfo(name = Constants.Quote.VOLUME_24H)
    private var volume24h: Double = Constants.Default.DOUBLE,
    @ColumnInfo(name = Constants.Quote.MARKET_CAP)
    private var marketCap: Double = Constants.Default.DOUBLE,
    @ColumnInfo(name = Constants.Quote.CHANGE_1H)
    private var change1h: Double = Constants.Default.DOUBLE,
    @ColumnInfo(name = Constants.Quote.CHANGE_24H)
    private var change24h: Double = Constants.Default.DOUBLE,
    @ColumnInfo(name = Constants.Quote.CHANGE_7D)
    private var change7d: Double = Constants.Default.DOUBLE,
    @ColumnInfo(name = Constants.Quote.LAST_UPDATED)
    private var lastUpdated: Long = Constants.Default.LONG
) : Base() {

    @Ignore
    constructor() : this(time = TimeUtilKt.currentMillis()) {

    }

    constructor(id: String) : this(time = TimeUtilKt.currentMillis(), id = id) {

    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val item = other as Quote
        return Objects.equal(item.id, id) && Objects.equal(item.currency, currency)
    }

    override fun hashCode(): Int {
        return Objects.hashCode(id, currency)
    }

    @PropertyName(Constants.Database.Crypto.VOLUME_24H)
    fun setVolume24h(volume24h: Double) {
        this.volume24h = volume24h
    }

    @PropertyName(Constants.Database.Crypto.VOLUME_24H)
    fun getVolume24h(): Double {
        return volume24h
    }

    @PropertyName(Constants.Database.Crypto.MARKET_CAP)
    fun setMarketCap(marketCap: Double) {
        this.marketCap = marketCap
    }

    @PropertyName(Constants.Database.Crypto.MARKET_CAP)
    fun getMarketCap(): Double {
        return marketCap
    }

    @PropertyName(Constants.Database.Crypto.CHANGE_1H)
    fun setChange1h(change1h: Double) {
        this.change1h = change1h
    }

    @PropertyName(Constants.Database.Crypto.CHANGE_1H)
    fun getChange1h(): Double {
        return change1h
    }

    @PropertyName(Constants.Database.Crypto.CHANGE_24H)
    fun setChange24h(change24h: Double) {
        this.change24h = change24h
    }

    @PropertyName(Constants.Database.Crypto.CHANGE_24H)
    fun getChange24h(): Double {
        return change24h
    }

    @PropertyName(Constants.Database.Crypto.CHANGE_7D)
    fun setChange7d(change7d: Double) {
        this.change7d = change7d
    }

    @PropertyName(Constants.Database.Crypto.CHANGE_7D)
    fun getChange7d(): Double {
        return change7d
    }

    @PropertyName(Constants.Database.Crypto.LAST_UPDATED)
    fun setLastUpdated(lastUpdated: Long) {
        this.lastUpdated = lastUpdated
    }

    @PropertyName(Constants.Database.Crypto.LAST_UPDATED)
    fun getLastUpdated(): Long {
        return lastUpdated
    }
}